### ISSUE TITLE
Version switcher without gh-pages

### ISSUE THREAD

**Issue by joestr** [xvik/gradle-mkdocs-plugin#31]
Is it possible to have a version switcher without requiring gh-pages? A versions.json must be present on the root but I tried such a setup but it wasn't successfull.

**Comment by joestr** [comment]
I gave it another try and noticed I just had a typo in the `versions.json`

To write a `versions.json` _by hand_ I only had to add following block in the `build.gradle` file:

```
task writeVersionsJson() {
  new File("$buildDir/mkdocs").mkdirs()
  new File("$buildDir/mkdocs", "versions.json").write("""
[
  {
    "version": "1.0.0",
    "title": "1.0.0",
    "aliases": [
      "latest"
    ]
  }
]
  """)
}
```

**Comment by xvik** [comment]
There is no strict github pages requirement - it could be any github repo (or branch in the current repo). There is a `mkdocsVersionsFile` task which generates versions file, but to do it it must know about the other (existing) versions. That's why it relies on git repository where all docs are published: checkout repo, collects all existing versions (folders) and generates file.

If you did not publish docs into git repo (branch) then your manual way is probably the best.

In theory, if all docs folder is reachable, versions generation could be automated (feature request candidate), but I need to know the exact use case.

**Comment by joestr** [comment]
Wouldn't it suffice just to reference to an older `versions.json` file *assuming* the folder structure is correct or does it strictly required the folders for checking?

**Comment by xvik** [comment]
In theory, previous versions file could also be used, if available. But I dont see overall use case (how versions file could be abvailable but not actual folders). Only if just generated docs always copied into some ftp (remote location), but then how project could know about actual vesrsions file - probably it assumed to be committed with the project. If you describe your case in more details, I could think about it.

I rely on real folders becuase it avoid mistakes: generated versions file would 100% contain only existing versions (and 100% all of them). Also, if some docs version was removed (e.g. previous snapshot folder removed) - it would be automtically removed from versions file on next puplication. But this is driven by my primary use case - github pages (so I can always have all versions).

**Comment by joestr** [comment]
To be more specific, my docs are deployed to a folder via FTP and the problem was that I didn't want to use gh-pages for it.
So I started tinkering around crafting the `versions.json` file by hand (which seems to work just fine).

To further elaborate my question: Could the `mkdocsVersionsFile` task be tuned (or a seperate task be created) which generates the `versions.json` file based on another `versions.json` file?

like `mkdocsVersionsFileFromPrevious`

```
mkdocs {
  publish {
    docPath = '1.0.0'
    rootRedirect = true
    rootRedirectTo = 'latest'
    versionAliases = ['latest']
    generateVersionsFile = true
    generateFromPrevious = true
    previousVersionsFile = 'https://www.example.com/owca-projects/a-cool-project/versions.json'
  }
}
```

Someone deleting a specific version from the docs root and deliberately not updating the `versions.json` _could_ cause problems but in my opinion this is neglectable.

**Comment by xvik** [comment]
I think one new property would be enough:

```
mkdocs {
  publish {
    docPath = '1.0.0'
    rootRedirect = true
    rootRedirectTo = 'latest'
    versionAliases = ['latest']
    generateVersionsFile = true
    currentVersionsFile = 'https://www.example.com/owca-projects/a-cool-project/versions.json'
  }
}
```

`currentVersionsFile` should support http and local files. 

If current file is specified then `mkdocsBuild` task would generate updated versions.json. If file not found, new empty versions.json would be created (if file specified then assuming incremental updates assumed and absence of file just means first generation). Existing option `generateVersionsFile = false` would prevent versions update/generation - so no need for additinoal options.
 
Publication workflow would remain the same (just would skip versions generation if its already exists).

For me, looks good enough. Can't promise to add this very soon.

**Comment by xvik** [comment]
Implemented in version [3.0.0](https://github.com/xvik/gradle-mkdocs-plugin/releases/tag/3.0.0). You can specify now path to existing versions file:

```groovy
mkdocs.publish.existingVersionsFile = 'https://xvik.github.io/gradle-mkdocs-plugin/versions.json'
```

And `mkdocsBuild` task would generate updated versions.json in mkdocs build dir. 
Note that task will ignore download errors - if it can't load remote file, new empty file would be created (in order to support furst publication case, when remote file does not yet exists).

Also, plugin was split into two plugins, so now you can use `ru.vyarus.mkdocs-build` plugin in order to avoid publication tasks registration

[Related docs](https://xvik.github.io/gradle-mkdocs-plugin/latest/guide/multi-version/#incremental-versions)
